be/tx_bulk.c

```diff
@@ -231,41 +415,106 @@ static void be_tx_bulk_close_cb(struct m0_sm_group *grp, struct m0_sm_ast *ast)
 	struct m0_be_tx_bulk_cfg  *tb_cfg;
 	struct be_tx_bulk_worker  *worker = ast->sa_datum;
 	struct m0_be_tx_bulk      *tb;
+	uint64_t                   i;
 
+	M0_ENTRY("worker=%p", worker);
 	M0_PRE(ast == &worker->tbw_close);
 	tb = worker->tbw_tb;
 	tb_cfg = &tb->btb_cfg;
-	M0_BE_OP_SYNC(op, tb_cfg->tbc_do(tb, &worker->tbw_tx, &op,
-	                                 tb_cfg->tbc_datum, worker->tbw_user));
+	for (i = 0; i < worker->tbw_items_nr; ++i) {
+		M0_BE_OP_SYNC(op, tb_cfg->tbc_do(tb, &worker->tbw_tx, &op,
+		                                 tb_cfg->tbc_datum,
+		                                 worker->tbw_item[i].bbd_user));
+	}
 	m0_be_tx_close(&worker->tbw_tx);
+	M0_LEAVE("worker=%p", worker);
 }
 
 static void be_tx_bulk_gc_cb(struct m0_be_tx *tx, void *param)
 {
 	struct be_tx_bulk_worker *worker = param;
+	struct m0_be_tx_bulk     *tb;
+	uint64_t                  i;
 
+	M0_ENTRY("worker=%p", worker);
 	M0_PRE(tx == &worker->tbw_tx);
 
-	m0_sm_ast_post(worker->tbw_grp, &worker->tbw_init);
+	tb = worker->tbw_tb;
+	for (i = 0; i < worker->tbw_items_nr; ++i) {
+		tb->btb_cfg.tbc_done(tb, tb->btb_cfg.tbc_datum,
```

Comment:
As discussed can we pass worker id (index) to call back function ? This will help application to identify worker and do processing accordingly.
 index = worker - &tb->btb_worker[0];